src/paradigma/heart_rate/heart_rate_analysis.py

```diff
@@ -85,4 +87,60 @@ def signal_quality_classification_io(input_path: Union[str, Path], output_path:
     metadata_time, metadata_values = read_metadata(input_path, config.meta_filename, config.time_filename, config.values_filename)
     df = tsdf.load_dataframe_from_binaries([metadata_time, metadata_values], tsdf.constants.ConcatenationType.columns)
 
-    df = signal_quality_classification(df, config, path_to_classifier_input)
\ No newline at end of file
+    df = signal_quality_classification(df, config, path_to_classifier_input)
+
+
+def estimate_heart_rate(df: pd.DataFrame, df_ppg_preprocessed: pd.DataFrame, config:HeartRateExtractionConfig) -> pd.DataFrame:  
+    """
+    Estimate the heart rate from the PPG signal using the time-frequency domain method.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The DataFrame containing the PPG signal.
+    df_ppg_preprocessed : pd.DataFrame
+        The DataFrame containing the preprocessed PPG signal.
+    config : HeartRateExtractionConfig
+        The configuration for the heart rate estimation.
+
+    Returns
+    -------
+    pd.DataFrame
+        The DataFrame containing the heart rate estimations.
+    """
+       
+    # Assign window-level probabilities to individual samples
+    ppg_post_prob = df.loc[:, DataColumns.PRED_SQA_PROBA].to_numpy()
+    #acc_label = df.loc[:, DataColumns.ACCELEROMETER_LABEL].to_numpy() # Adjust later in data columns to get the correct label, should be first intergrated in feature extraction and classification
+
+    sqa_label = assign_sqa_label(ppg_post_prob, config)
+    v_start_idx, v_end_idx = extract_hr_segments(sqa_label, config.min_hr_samples)
+    
+    fs = config.sampling_frequency
+
+    v_hr_rel = np.array([])
+    t_hr_rel = np.array([])
+
+    for start_idx, end_idx in zip(v_start_idx, v_end_idx):
+        # Skip if the epoch cannot be extended by 2s on both sides
+        if start_idx < 2 * fs or end_idx > len(df_ppg_preprocessed) - 2 * fs:
```

Comment:
You seem to be using `2 * fs` quite often, perhaps you can assign this a variable once.